// mobx-react-form/lib/Base.js
"use strict";

function _typeof(obj) { "@babel/helpers - typeof"; if (typeof Symbol === "function" && typeof Symbol.iterator === "symbol") { _typeof = function _typeof(obj) { return typeof obj; }; } else { _typeof = function _typeof(obj) { return obj && typeof Symbol === "function" && obj.constructor === Symbol && obj !== Symbol.prototype ? "symbol" : typeof obj; }; } return _typeof(obj); }

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports["default"] = void 0;

var _isFunction2 = _interopRequireDefault(require("lodash/isFunction"));

var _mobx = require("mobx");

var _Actions2 = _interopRequireDefault(require("./shared/Actions"));

var _utils = require("./utils");

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { "default": obj }; }

function ownKeys(object, enumerableOnly) { var keys = Object.keys(object); if (Object.getOwnPropertySymbols) { var symbols = Object.getOwnPropertySymbols(object); if (enumerableOnly) symbols = symbols.filter(function (sym) { return Object.getOwnPropertyDescriptor(object, sym).enumerable; }); keys.push.apply(keys, symbols); } return keys; }

function _objectSpread(target) { for (var i = 1; i < arguments.length; i++) { var source = arguments[i] != null ? arguments[i] : {}; if (i % 2) { ownKeys(Object(source), true).forEach(function (key) { _defineProperty(target, key, source[key]); }); } else if (Object.getOwnPropertyDescriptors) { Object.defineProperties(target, Object.getOwnPropertyDescriptors(source)); } else { ownKeys(Object(source)).forEach(function (key) { Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key)); }); } } return target; }

function _toConsumableArray(arr) { return _arrayWithoutHoles(arr) || _iterableToArray(arr) || _unsupportedIterableToArray(arr) || _nonIterableSpread(); }

function _nonIterableSpread() { throw new TypeError("Invalid attempt to spread non-iterable instance.\nIn order to be iterable, non-array objects must have a [Symbol.iterator]() method."); }

function _unsupportedIterableToArray(o, minLen) { if (!o) return; if (typeof o === "string") return _arrayLikeToArray(o, minLen); var n = Object.prototype.toString.call(o).slice(8, -1); if (n === "Object" && o.constructor) n = o.constructor.name; if (n === "Map" || n === "Set") return Array.from(o); if (n === "Arguments" || /^(?:Ui|I)nt(?:8|16|32)(?:Clamped)?Array$/.test(n)) return _arrayLikeToArray(o, minLen); }

function _iterableToArray(iter) { if (typeof Symbol !== "undefined" && Symbol.iterator in Object(iter)) return Array.from(iter); }

function _arrayWithoutHoles(arr) { if (Array.isArray(arr)) return _arrayLikeToArray(arr); }

function _arrayLikeToArray(arr, len) { if (len == null || len > arr.length) len = arr.length; for (var i = 0, arr2 = new Array(len); i < len; i++) { arr2[i] = arr[i]; } return arr2; }

function _classCallCheck(instance, Constructor) { if (!(instance instanceof Constructor)) { throw new TypeError("Cannot call a class as a function"); } }

function _defineProperties(target, props) { for (var i = 0; i < props.length; i++) { var descriptor = props[i]; descriptor.enumerable = descriptor.enumerable || false; descriptor.configurable = true; if ("value" in descriptor) descriptor.writable = true; Object.defineProperty(target, descriptor.key, descriptor); } }

function _createClass(Constructor, protoProps, staticProps) { if (protoProps) _defineProperties(Constructor.prototype, protoProps); if (staticProps) _defineProperties(Constructor, staticProps); return Constructor; }

function _inherits(subClass, superClass) { if (typeof superClass !== "function" && superClass !== null) { throw new TypeError("Super expression must either be null or a function"); } subClass.prototype = Object.create(superClass && superClass.prototype, { constructor: { value: subClass, writable: true, configurable: true } }); if (superClass) _setPrototypeOf(subClass, superClass); }

function _setPrototypeOf(o, p) { _setPrototypeOf = Object.setPrototypeOf || function _setPrototypeOf(o, p) { o.__proto__ = p; return o; }; return _setPrototypeOf(o, p); }

function _createSuper(Derived) { var hasNativeReflectConstruct = _isNativeReflectConstruct(); return function _createSuperInternal() { var Super = _getPrototypeOf(Derived), result; if (hasNativeReflectConstruct) { var NewTarget = _getPrototypeOf(this).constructor; result = Reflect.construct(Super, arguments, NewTarget); } else { result = Super.apply(this, arguments); } return _possibleConstructorReturn(this, result); }; }

function _possibleConstructorReturn(self, call) { if (call && (_typeof(call) === "object" || typeof call === "function")) { return call; } return _assertThisInitialized(self); }

function _assertThisInitialized(self) { if (self === void 0) { throw new ReferenceError("this hasn't been initialised - super() hasn't been called"); } return self; }

function _isNativeReflectConstruct() { if (typeof Reflect === "undefined" || !Reflect.construct) return false; if (Reflect.construct.sham) return false; if (typeof Proxy === "function") return true; try { Date.prototype.toString.call(Reflect.construct(Date, [], function () {})); return true; } catch (e) { return false; } }

function _getPrototypeOf(o) { _getPrototypeOf = Object.setPrototypeOf ? Object.getPrototypeOf : function _getPrototypeOf(o) { return o.__proto__ || Object.getPrototypeOf(o); }; return _getPrototypeOf(o); }

function _defineProperty(obj, key, value) { if (key in obj) { Object.defineProperty(obj, key, { value: value, enumerable: true, configurable: true, writable: true }); } else { obj[key] = value; } return obj; }

var Base = /*#__PURE__*/function (_Actions) {
  _inherits(Base, _Actions);

  var _super = _createSuper(Base);

  function Base() {
    var _this;

    _classCallCheck(this, Base);

    _this = _super.call(this);

    _defineProperty(_assertThisInitialized(_this), "noop", function () {});

    _defineProperty(_assertThisInitialized(_this), "$submitted", 0);

    _defineProperty(_assertThisInitialized(_this), "$submitting", false);

    _defineProperty(_assertThisInitialized(_this), "$validated", 0);

    _defineProperty(_assertThisInitialized(_this), "$validating", false);

    _defineProperty(_assertThisInitialized(_this), "execHook", function (name) {
      var fallback = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
      return (0, _utils.$try)(fallback[name], _this.$hooks[name], _this.hooks && _this.hooks.apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)])[name], _this.noop).apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)]);
    });

    _defineProperty(_assertThisInitialized(_this), "execHandler", function (name, args) {
      var fallback = arguments.length > 2 && arguments[2] !== undefined ? arguments[2] : null;
      return [(0, _utils.$try)(_this.$handlers[name] && _this.$handlers[name].apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)]), _this.handlers && _this.handlers.apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)])[name] && _this.handlers.apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)])[name].apply(_assertThisInitialized(_this), [_assertThisInitialized(_this)]), fallback, _this.noop).apply(_assertThisInitialized(_this), _toConsumableArray(args)), _this.execHook(name)];
    });

    _defineProperty(_assertThisInitialized(_this), "intercept", function (opt) {
      return _this.MOBXEvent((0, _isFunction2["default"])(opt) ? {
        type: 'interceptor',
        call: opt
      } : _objectSpread({
        type: 'interceptor'
      }, opt));
    });

    _defineProperty(_assertThisInitialized(_this), "observe", function (opt) {
      return _this.MOBXEvent((0, _isFunction2["default"])(opt) ? {
        type: 'observer',
        call: opt
      } : _objectSpread({
        type: 'observer'
      }, opt));
    });

    _defineProperty(_assertThisInitialized(_this), "onClear", function () {
      for (var _len = arguments.length, args = new Array(_len), _key = 0; _key < _len; _key++) {
        args[_key] = arguments[_key];
      }

      return _this.execHandler('onClear', args, function (e) {
        e.preventDefault();

        _this.clear(true);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onReset", function () {
      for (var _len2 = arguments.length, args = new Array(_len2), _key2 = 0; _key2 < _len2; _key2++) {
        args[_key2] = arguments[_key2];
      }

      return _this.execHandler('onReset', args, function (e) {
        e.preventDefault();

        _this.reset(true);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onSubmit", function () {
      for (var _len3 = arguments.length, args = new Array(_len3), _key3 = 0; _key3 < _len3; _key3++) {
        args[_key3] = arguments[_key3];
      }

      return _this.execHandler('onSubmit', args, function (e) {
        var o = arguments.length > 1 && arguments[1] !== undefined ? arguments[1] : {};
        e.preventDefault();

        _this.submit(o);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onAdd", function () {
      for (var _len4 = arguments.length, args = new Array(_len4), _key4 = 0; _key4 < _len4; _key4++) {
        args[_key4] = arguments[_key4];
      }

      return _this.execHandler('onAdd', args, function (e, val) {
        e.preventDefault();

        _this.add((0, _utils.$isEvent)(val) ? null : val);
      });
    });

    _defineProperty(_assertThisInitialized(_this), "onDel", function () {
      for (var _len5 = arguments.length, args = new Array(_len5), _key5 = 0; _key5 < _len5; _key5++) {
        args[_key5] = arguments[_key5];
      }

      return _this.execHandler('onDel', args, function (e, path) {
        e.preventDefault();

        _this.del((0, _utils.$isEvent)(path) ? _this.path : path);
      });
    });

    (0, _mobx.makeObservable)(_assertThisInitialized(_this), {
      $submitted: _mobx.observable,
      $submitting: _mobx.observable,
      $validated: _mobx.observable,
      $validating: _mobx.observable,
      submitted: _mobx.computed,
      submitting: _mobx.computed,
      validated: _mobx.computed,
      validating: _mobx.computed,
      hasIncrementalKeys: _mobx.computed,
      hasNestedFields: _mobx.computed,
      size: _mobx.computed
    });
    return _this;
  }

  _createClass(Base, [{
    key: "submitted",
    get: function get() {
      return (0, _mobx.toJS)(this.$submitted);
    }
  }, {
    key: "submitting",
    get: function get() {
      return (0, _mobx.toJS)(this.$submitting);
    }
  }, {
    key: "validated",
    get: function get() {
      return (0, _mobx.toJS)(this.$validated);
    }
  }, {
    key: "validating",
    get: function get() {
      return (0, _mobx.toJS)(this.$validating);
    }
  }, {
    key: "hasIncrementalKeys",
    get: function get() {
      return this.fields.size && (0, _utils.hasIntKeys)(this.fields);
    }
  }, {
    key: "hasNestedFields",
    get: function get() {
      return this.fields.size !== 0;
    }
  }, {
    key: "size",
    get: function get() {
      return this.fields.size;
    }
    /**
     Interceptor
     */

  }]);

  return Base;
}(_Actions2["default"]);

exports["default"] = Base;
module.exports = exports["default"];
